pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,437 @@
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/gogo/protobuf/proto"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	mqtt_pubsub "github.com/mainflux/mainflux/pkg/messaging/mqtt"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	topic       = "topic"
+	chansPrefix = "channels"
+	channel     = "9b7b1b3f-b1b0-46a8-a717-b8213f9eda3b"
+	subtopic    = "engine"
+)
+
+var (
+	data = []byte("payload")
+)
+
+func TestPublisher(t *testing.T) {
+	msgChan := make(chan []byte)
+
+	// Subscribing with topic, and with subtopic, so that we can publish messages
+	client, err := newClient(address, "clientID1", 30*time.Second)
+	assert.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+
+	token := client.Subscribe(topic, qos, func(c mqtt.Client, m mqtt.Message) {
+		msgChan <- m.Payload()
+	})
+	if ok := token.WaitTimeout(100 * time.Millisecond); !ok {
+		assert.Fail(t, fmt.Sprintf("failed to subscribe to topic %s", topic))
+	}
+	assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+
+	token = client.Subscribe(fmt.Sprintf("%s.%s", topic, subtopic), qos, func(c mqtt.Client, m mqtt.Message) {
+		msgChan <- m.Payload()
+	})
+	if ok := token.WaitTimeout(100 * time.Millisecond); !ok {
+		assert.Fail(t, fmt.Sprintf("failed to subscribe to topic %s", topic))
+	}
+	assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+
+	t.Cleanup(func() {
+		token := client.Unsubscribe(topic, fmt.Sprintf("%s.%s", topic, subtopic))
+		token.WaitTimeout(100 * time.Millisecond)
+		assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+
+		client.Disconnect(100)
+	})
+
+	// publish with empty topic
+	err = pubsub.Publish("", messaging.Message{Payload: data})
+	assert.Equal(t, err, mqtt_pubsub.ErrEmptyTopic, fmt.Sprintf("Publish with empty topic: expected: %s, got: %s", mqtt_pubsub.ErrEmptyTopic, err))
+
+	cases := []struct {
+		desc     string
+		channel  string
+		subtopic string
+		payload  []byte
+	}{
+		{
+			desc:    "publish message with nil payload",
+			payload: nil,
+		},
+		{
+			desc:    "publish message with string payload",
+			payload: data,
+		},
+		{
+			desc:    "publish message with channel",
+			payload: data,
+			channel: channel,
+		},
+		{
+			desc:     "publish message with subtopic",
+			payload:  data,
+			subtopic: subtopic,
+		},
+		{
+			desc:     "publish message with channel and subtopic",
+			payload:  data,
+			channel:  channel,
+			subtopic: subtopic,
+		},
+	}
+	for _, tc := range cases {
+		expectedMsg := messaging.Message{
+			Publisher: "clientID11",
+			Channel:   tc.channel,
+			Subtopic:  tc.subtopic,
+			Payload:   tc.payload,
+		}
+		err := pubsub.Publish(topic, expectedMsg)
+		assert.Nil(t, err, fmt.Sprintf("%s: got unexpected error: %s\n", tc.desc, err))
+
+		data, err := proto.Marshal(&expectedMsg)
+		assert.Nil(t, err, fmt.Sprintf("%s: got unexpected error: %s\n", tc.desc, err))
```

Comment:
Use `failed to serialize protobuf error:` instead of generic `got unexpected error:`